### PR TITLE
2fa is here

### PR DESCRIPTION
**Pull request by eprei**
**New feature available** 🎉

From now on, the user can activate the use of 2fa directly from his profile page.

When the user clicks to activate it, he is redirected to the page localhost:4040/tfaturnOn . There he is provided with a QR code that he must scan with a TOTP code generation application and then send the verification code it generated to our server. If the code is correct, the user's TFAEnabled attribute is changed to true and will be taken into account the next time he/she tries to log in.

The page localhost:4040/tfaturnOn should ideally not be accessible if the user has already enabled 2fa so if you have ideas on how to implement it, they are welcome.

Most of our endpoints will be protected by the 'AuthenticatedGuard' guard. So now AuthGuard checks two things. Firstly, that the user has successfully logged in with 42. Secondly, it verifies that the user has successfully executed the 2fa (if in its attributes in the database TFAEnabled === true ).

There is also the page localhost:4040/tfaverify . This page should be shown to a user who (having previously and successfully activated 2fa) has successfully logged in with his login 42. (If you have ideas how to make these redirections, they are welcome).
This page exists for the user to send his verification code to the backend for verification and if it is correct, he is allowed to login to his profile.
This page should not be accessible once the user has entered their code correctly. From the backend, the endpoint related to this page is already protected taking into account the last consideration but not yet in the frontend (if you have ideas on how to implement it, they are welcome).

Note that the creation of the user in our database (once the user successfully logs in with 42) is still random. Each time we login with 42 a random user is created. This feature is being worked on in s_create_player.

To test the new feature you can download your favorite TOTP creation application and try to enable and disable 2fa.

**Documentation and sources:**

* [NestJS Authentication: JWTs, Sessions, logins, and more! | NestJS PassportJS Tutorial](https://youtu.be/_L225zpUK0M)
 
* [NestJS Guards](https://docs.nestjs.com/guards#guards)

* [Passport](https://www.passportjs.org/)

* [Passport Strategies](https://www.passportjs.org/concepts/authentication/strategies/)

* [Passport Sessions](https://www.passportjs.org/concepts/authentication/sessions/)

* [Exoress-session](https://github.com/expressjs/session)

* [Express-session: how it works](https://github.com/rohan-paul/Awesome-JavaScript-Interviews/blob/master/Node-Express/passport-express-session-how-it-works.md)

* [Understanding Sessions and Local Authentication in Express with Passport and MongoDb](
https://mianlabs.com/2018/05/09/understanding-sessions-and-local-authentication-in-express-with-passport-and-mongodb/)

* [2FA with NestJs & passeport using Google Authenticator](https://dev.to/hahnmatthieu/2fa-with-nestjs-passeport-using-google-authenticator-1l32)

* [Otplib](https://github.com/yeojz/otplib#about)
  Otplib is a JavaScript One Time Password (OTP) library for OTP generation and verification that we will use in 2fa.

* [Qrcode](https://github.com/soldair/node-qrcode)
  Qrcode is a barcode and qrcode generator that we will use to transform the otpAuthUrl (generated by the otplib library) into qrcode to be later scanned by the authentication google application.

### CODE REVIEW

File: backend/src/auth/auth.service.ts [hunk 1]
```diff
@@ -1,4 +1,120 @@
...
+        if (!user) {
+            throw new NotFoundException('User not found')
+        }
+        const secretAndUrl = await this.generateTwoFactorAuthenticationSecret(
+            user
+        )
+
+        const qrCode = await this.generateQrCodeDataURL(
+            (
+                await secretAndUrl
```
Comment (RobinBurri) [inline backend/src/auth/auth.service.ts:115]: secretAndUrl is not a Promise, so no await.

File: backend/src/auth/guards/totp.guard.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common'
```
Comment (RobinBurri) [inline backend/src/auth/guards/totp.guard.ts:1]: I cannot find this file. WTF?

**Review by RobinBurri — COMMENTED** [review]
🚀

File: backend/src/auth/guards/totp.guard.ts (hunk 2 above)
Comment (eprei) [reply backend/src/auth/guards/totp.guard.ts:1]: Did you find it?

File: backend/src/auth/auth.service.ts (hunk 1 above)
Comment (eprei) [reply backend/src/auth/auth.service.ts]: This way is it ok? 940a4a008d5990e88911a836a1c732710492a136

File: backend/src/auth/guards/totp.guard.ts (hunk 2 above)
Comment (RobinBurri) [reply backend/src/auth/guards/totp.guard.ts:1]: yes

File: backend/src/auth/auth.service.ts [hunk 3]
```diff
@@ -1,4 +1,121 @@
...
+import { UserService } from 'src/user/user.service'
+import { toDataURL } from 'qrcode'
 
 @Injectable()
-export class AuthService {}
+export class AuthService {
+    constructor(private userService: UserService) {}
+
+    async generateTwoFactorAuthenticationSecret(user: User) {
+        const secret = authenticator.generateSecret()
```
Comment (RobinBurri) [inline backend/src/auth/auth.service.ts:18]:
```suggestion
        const secret : string = authenticator.generateSecret()
```

Add some type for clearity

File: backend/src/auth/auth.service.ts [hunk 4]
```diff
@@ -1,4 +1,121 @@
...
+
+        const otpauthUrl = authenticator.keyuri(
```
Comment (RobinBurri) [inline backend/src/auth/auth.service.ts:20]:
```suggestion
        const otpauthUrl: string = authenticator.keyuri(
```

same here

File: backend/src/auth/auth.service.ts [hunk 5]
```diff
@@ -1,4 +1,121 @@
...
+    }
+
+    async generateQrCodeDataURL(otpAuthUrl: string) {
+        return toDataURL(otpAuthUrl)
+    }
+
+    isTwoFactorAuthenticationCodeValid(
+        twoFactorAuthenticationCode: string,
+        User: User
+    ) {
```
Comment (RobinBurri) [inline backend/src/auth/auth.service.ts:46]:
```suggestion
    ): boolean {
```
same here

**Review by RobinBurri — COMMENTED** [review]
🦺

**Review by RobinBurri — APPROVED** [review]
👍

**Comment by eprei** [comment]
@fire-poy Don't worry, we will soon find the solution. You can also validate the PR without the need to fix this issue as you have done.

**Review by shaolin-peanut — APPROVED** [review]
Congrats for putting all of this together!

**Comment by t-h2o** [comment]
Should we generate the QR code client side ?

**Pull request merged**